iot-e2e-tests/common/src/test/java/tests/integration/com/microsoft/azure/sdk/iot/iothub/twin/TwinPnPTests.java

```diff
@@ -0,0 +1,185 @@
+    {
+        this.testInstance = new TwinPnPTestInstance(protocol);
+    }
+
+    public class TwinPnPTestInstance
+    {
+        public IotHubClientProtocol protocol;
+        private Device deviceForRegistryManager;
+        private Module moduleForRegistryManager;
+
+        private DeviceTwin sCDeviceTwin;
+        private TwinPnPTests.TestDevice testDevice;
+
+        public TwinPnPTestInstance(IotHubClientProtocol protocol) throws IOException
+        {
+            this.protocol = protocol;
+            this.sCDeviceTwin = DeviceTwin.createFromConnectionString(iotHubConnectionString);
+        }
+    }
+
+    @AfterClass
+    public static void tearDown() throws Exception
+    {
+        if (registryManager != null)
+        {
+            registryManager.close();
+            registryManager = null;
+        }
+    }
+
+    private void createDeviceWithClientOptions(IotHubClientProtocol protocol, ClientOptions options) throws IOException, URISyntaxException, InterruptedException
+    {
+        testInstance.testDevice.deviceClient = new DeviceClient(DeviceConnectionString.get(iotHubConnectionString, testInstance.deviceForRegistryManager), protocol, options);
+        testInstance.testDevice.deviceClient.open();
+    }
+
+    private void createModuleWithClientOptions(IotHubClientProtocol protocol, ClientOptions options) throws IOException, URISyntaxException, InterruptedException, ModuleClientException {
+        testInstance.testDevice.moduleClient = new ModuleClient(DeviceConnectionString.get(iotHubConnectionString, testInstance.deviceForRegistryManager, testInstance.moduleForRegistryManager), protocol, options);
+        testInstance.testDevice.moduleClient.open();
+    }
+
+    @Before
+    public void createDevice() throws Exception
+    {
+        testInstance.testDevice = new TestDevice();
+        testInstance.testDevice.deviceId = "java-twin-PnP-e2e-test-".concat(UUID.randomUUID().toString());
+        testInstance.testDevice.moduleId = "java-twin-PnP-module-e2e-test-".concat(UUID.randomUUID().toString());
+
+        testInstance.deviceForRegistryManager = com.microsoft.azure.sdk.iot.service.Device.createFromId(testInstance.testDevice.deviceId, null, null);
+        testInstance.deviceForRegistryManager = Tools.addDeviceWithRetry(registryManager, testInstance.deviceForRegistryManager);
+        testInstance.moduleForRegistryManager = com.microsoft.azure.sdk.iot.service.Module.createFromId(testInstance.testDevice.deviceId, testInstance.testDevice.moduleId,null);
+        testInstance.moduleForRegistryManager = Tools.addModuleWithRetry(registryManager, testInstance.moduleForRegistryManager);
+
+    }
+
+    @After
+    public void destroyDevice() throws Exception
+    {
+        if (testInstance.testDevice.deviceClient != null)
+        {
+            testInstance.testDevice.deviceClient.closeNow();
+            testInstance.testDevice.deviceClient = null;
```

Comment:
No need to set fields to null when you are done with them. Here and below

Reply:
why would you want a dangling object?

Reply:
http://net-informations.com/java/cjava/objectonull.htm

That doesn't help garbage collection as much as you might think